src/implementation.jl

```diff
@@ -0,0 +1,74 @@
+module Implementation
+
+import Base: isvatuple
+
+check(T) = (T===Tuple || T===NTuple) && throw(ArgumentError("parameters of $T are undefined"))
+
+# Define methods for up to N tuple parameters to be evaluated in the Type domain
```

Comment:
Hah, yeah, I suppose, but I have a feeling that allowing inference to reason about these things will be a huge gain.  See, for example:

```jl
julia> @code_typed TupleTypes.concatenate(Tuple{Int8,Int16}, Tuple{Int32, Int64})
1-element Array{Any,1}:
 :($(Expr(:lambda, Any[symbol("#s3"),symbol("#s2")], Any[Any[Any[symbol("#s3"),Type{Tuple{Int8,Int16}},0],Any[symbol("#s2"),Type{Tuple{Int32,Int64}},0]],Any[],Any[],Any[:P1,:P2,:Q1,:Q2]], :(begin  # /Users/mbauman/.julia/v0.4/TupleTypes/src/implementation.jl, line 34:
        return Tuple{Int8,Int16,Int32,Int64}
    end::Type{Tuple{Int8,Int16,Int32,Int64}}))))
```